pkg/porter/explain.go

```diff
@@ -0,0 +1,291 @@
+	fmt.Fprintf(p.Out, "Version: %s\n", bun.Version)
+	fmt.Fprintln(p.Out, "")
+
+	p.generateCredentialsBlock(bun)
+
+	p.generateParametersBlock(bun)
+
+	p.generateOutputsBlock(bun)
+
+	p.generateActionsBlock(bun)
+
+	return nil
+}
+
+func (p *Porter) generateCredentialsBlock(bun *PrintableBundle) {
+	if len(bun.Credentials) > 0 {
+		fmt.Fprintln(p.Out, "Credentials:")
+		p.generateCredentialsTable(bun)
+	} else {
+		fmt.Fprintln(p.Out, "No credentials defined")
+	}
+	fmt.Fprintln(p.Out, "") // force a blank line after this block
+}
+func (p *Porter) generateCredentialsTable(bun *PrintableBundle) {
+	h := []string{"Name", "Description", "Required"}
+	d := [][]string{}
+	for k, v := range bun.Credentials {
+		d = append(d, []string{k, v.Description, fmt.Sprintf("%v", v.Required)})
+	}
+	p.generateTable(h, d)
+}
+
+func (p *Porter) generateParametersBlock(bun *PrintableBundle) {
+	if len(bun.Parameters) > 0 {
+		fmt.Fprintln(p.Out, "Parameters:")
+		p.generateParametersTable(bun)
+	} else {
+		fmt.Fprintln(p.Out, "No parameters defined")
+	}
+	fmt.Fprintln(p.Out, "") // force a blank line after this block
+}
+func (p *Porter) generateParametersTable(bun *PrintableBundle) {
+	h := []string{"Name", "Description", "Type", "Default", "Required", "Applies To"}
+	d := [][]string{}
+	for k, v := range bun.Parameters {
+		d = append(d, []string{
+			k, v.Description,
+			fmt.Sprintf("%v", v.Type),
+			fmt.Sprintf("%v", v.Default),
+			fmt.Sprintf("%v", v.Required),
+			v.ApplyTo,
+		})
+	}
+	p.generateTable(h, d)
+}
+
+func (p *Porter) generateOutputsBlock(bun *PrintableBundle) {
+	if len(bun.Outputs) > 0 {
+		fmt.Fprintln(p.Out, "Outputs:")
+		p.generateOutputsTable(bun)
+	} else {
+		fmt.Fprintln(p.Out, "No outputs defined")
+	}
+	fmt.Fprintln(p.Out, "") // force a blank line after this block
+}
+
+func (p *Porter) generateOutputsTable(bun *PrintableBundle) {
+	h := []string{"Name", "Description", "Type", "Applies To"}
+	d := [][]string{}
+	for k, v := range bun.Outputs {
+		d = append(d, []string{k, v.Description, fmt.Sprintf("%v", v.Type), v.ApplyTo})
+	}
+	p.generateTable(h, d)
+}
+
+func generateApplyToString(appliesTo []string) string {
+	if len(appliesTo) == 0 {
+		return "All Actions"
+	} else {
+		return strings.Join(appliesTo, ",")
+	}
+}
+
+func (p *Porter) generateActionsBlock(bun *PrintableBundle) {
+	if len(bun.Actions) > 0 {
+		fmt.Fprintln(p.Out, "Actions:")
+		p.generateActionsTable(bun)
+	} else {
+		fmt.Fprintln(p.Out, "No custom actions defined")
+	}
+	fmt.Fprintln(p.Out, "") // force a blank line after this block
+}
+
+func (p *Porter) generateActionsTable(bun *PrintableBundle) {
+	h := []string{"Name", "Description", "Modifies Instance", "Stateless"}
+	d := [][]string{}
+	for k, v := range bun.Actions {
+		d = append(d, []string{k, v.Description, fmt.Sprintf("%v", v.Modifies), fmt.Sprintf("%v", v.Stateless)})
+	}
+	p.generateTable(h, d)
+}
+
+func (p *Porter) generateTable(h []string, d [][]string) {
```

Comment:
I'd prefer that we use the same table look, and function for printing tabular data throughout Porter. It'll help a ton as people will be copy/pasting this code when they tackle new issues.

Is there a reason why you didn't use pkg/printer.PrintTable? Each function `generate*` could have called it and then we didn't need a new table printing function. 

e.g.

https://github.com/deislabs/porter/blob/7da5cffdf3581c27dfab03af88df2df005baa02d/pkg/porter/mixins.go#L23-L32

Reply:
Oops, didn't see that. I can go refactor.